src/components/InputWithHightlights.vue

```diff
@@ -0,0 +1,140 @@
+<template>
+  <div>
+    <div class="selected__container" :style="inputStyle">
+      <div
+        v-for="(part, index) in parts"
+        :key="index" class="selected__line"
+        :style="{ marginLeft: -inputScrolledLeft + 'px' }"
+      >
+        <span :style="{whiteSpace: 'pre'}">{{ part.textBefore }}</span>
+
+        <span
+          :class="['selected', `selected--color-${part.color}`]"
+          :style="{
+            whiteSpace: 'pre',
+          }"
+        >{{ part.text }}</span>
+      </div>
+    </div>
+
+    <unnnic-input
+      ref="input"
+      :value="value"
+      @input="$emit('input', $event)"
+      v-bind="$attrs"
+    />
+  </div>
+</template>
+
+<script>
+import { getEntityColor } from '../utils/entitiesColors';
+
+export default {
+  props: {
+    value: String,
+    selected: Object,
+    entities: Array,
+  },
+
+  data() {
+    return {
+      inputStyle: {},
+      inputScrolledLeft: 0,
+    };
+  },
+
+  mounted() {
+    const input = this.$refs.input.$el.querySelector('input');
+
+    const style = getComputedStyle(input);
+
+    this.inputStyle = {
+      font: style.font,
+      fontWeight: style.fontWeight,
+      height: style.height,
+      marginBottom: `-${style.height}`,
+      padding: style.padding,
+      backgroundColor: style.backgroundColor,
+    }
+
+    input.style.backgroundColor = 'transparent';
+
+    input.addEventListener('scroll', (event) => {
+      this.inputScrolledLeft = event.target.scrollLeft;
+    });
+
+    input.addEventListener('select', (event) => {
+      const selected = event.target.value.substring(
+        event.target.selectionStart,
+        event.target.selectionEnd
+      );
+
+      const offsetLeft = selected.match(/^( +)[^ ]/)?.[1].length || 0;
+      const offsetRight = selected.match(/[^ ]( +)$/)?.[1].length || 0;
+
+      this.$emit('update:selected', {
+        start: event.target.selectionStart + offsetLeft,
+        end: event.target.selectionEnd - offsetRight,
+      });
+    });
+  },
+
+  computed: {
+    parts() {
+      return [].concat(this.selected ? {
+        textBefore: this.value.substring(0, this.selected.start),
+        text: this.value.substring(this.selected.start, this.selected.end),
+      } : []).concat(this.entities.map((entity) => ({
+        textBefore: this.value.substring(0, entity.start),
+        text: this.value.substring(entity.start, entity.end),
+        color: getEntityColor(entity.entity),
+      })));
```

Comment:
It's difficult and tiring to read this code, try using variables to describe it better.